Pecos hybrid axibump runs

clarkpede opened the pull request:
This is PR 4 of a series of PRs intended to get the pecos-dev branch back in line with the "best" and most stable code we are currently using.

clarkpede committed:
Cherry-pick 8432301

Add a switch to change the v2f realizability limit

There's now two options: to use the realizability limits on T and L, as
was previously done, or to only limit the eddy viscosity.

Conflicts:
        Common/include/config_structure.hpp
	Common/include/config_structure.inl
	Common/src/config_structure.cpp
	SU2_CFD/src/solver_direct_turbulent_v2f.cpp
	config_template.cfg

trevilo committed:
Towards getting axibump hybrid to run, including axi-periodic forcing and hacks trying to assess divergence

trevilo committed:
Communication and bug fixes for resolution tensor

trevilo committed:
fix due to resolution tensor changes

trevilo committed:
modify rotationally periodic version of forcing

trevilo committed:
Use mean density in M43, and modify AR switch slightly

trevilo committed:
bulk viscosity (in comments)

trevilo committed:
force production of v2 to be positive (for stability)

trevilo committed:
turn min dissipation back down to 0.01

trevilo committed:
fix transformation for tensors (velocity gradients) on rotationally periodic boundaries

trevilo committed:
don't use stagnation point anomaly time scale in averaging

trevilo committed:
be more chatty when non-physical density/temperature/etc encountered

clarkpede committed:
Merge branch 'pecos-dev' into pecos-hybrid-axibump-runs

clarkpede commented on Common/src/geometry_structure.cpp:
```diff
@@ -12018,13 +12019,18 @@ void CGeometry::SetResolutionTensor(void) {
       for (unsigned short iDim = 0; iDim < nDim; iDim++) {
         for (unsigned short jDim = 0; jDim < nDim; jDim++) {
           su2double temp_value = temp_tensor[iDim][jDim] / (node[iPoint]->GetnElem());
-          if (temp_value > EPS)
```
The "positivity" test was originally included to make sure we stayed above machine precision.  Otherwise, I would get funny numbers (that were still technically correct) like 1.0000000000000001 or 0.99999999999999999.

So this change is fine by me.

trevilo replied:
Only adding the positive contributions is incorrect.  That is the reason for the change.  The eigenvalues of the resolution tensor have to be positive, but the entries do not, so if you only add the positive entries here you'll get the wrong answer (on a general mesh... the result on an orthogonal mesh should be ok).  If you want to avoid adding values that are order machine precision (e.g., to avoid the funny numbers you mention) you could test that abs(temp_value) > EPS, but I don't see the point.  That is what the cryptic comment was intended to convey.

clarkpede commented on SU2_CFD/src/numerics_direct_mean.cpp:
```diff
@@ -2964,6 +2964,13 @@ void CUpwRoe_Flow::ComputeResidual(su2double *val_residual, su2double **val_Jaco
+  dissipationVec[2] = Dissipation_ij;
+  dissipationVec[3] = Dissipation_ij; //max(Dissipation_ij, 0.1);
+  dissipationVec[4] = Dissipation_ij; //max(Dissipation_ij, 0.1);
+
```
What's going on here? I'm guessing you were trying different central/upwind weightings for different flow equations?

trevilo replied:
Yes, you're right, but it did not solve the problem I was trying to get at, so we can take it out.  At the moment it is just clutter.

clarkpede commented on SU2_CFD/src/numerics_direct_mean.cpp:
```diff
@@ -2975,12 +2982,23 @@ void CUpwRoe_Flow::ComputeResidual(su2double *val_residual, su2double **val_Jaco
+        //   // val_Jacobian_j[iVar][jVar] -= (1.0-kappa)*Proj_ModJac_Tensor_ij*Area;
+        //   val_Jacobian_i[iVar][jVar] += (1.0-kappa)*Proj_ModJac_Tensor_ij*Area*Dissipation_ij;
+        //   val_Jacobian_j[iVar][jVar] -= (1.0-kappa)*Proj_ModJac_Tensor_ij*Area*Dissipation_ij;
+
```
Should we put the `Dissipation_ij` back in the Jacobian?

trevilo replied:
Yes, the behavior here should be the same as it was previously, since Dissipation_ij goes into dissipationVec, which is included in Proj_ModJac_Tensor_ij.  But, I vote we revert these changes since the behavior should be the same and the previous code was simpler.

clarkpede commented on SU2_CFD/src/numerics_direct_mean.cpp:
```diff
@@ -4576,11 +4594,18 @@ void CAvgGrad_Base::SetTauJacobian(const su2double *val_Mean_PrimVar,
+  //const su2double xiBulk = 20.0*total_viscosity/(Density*val_dist_ij);
+  //const su2double xiBulk = 100.0*total_viscosity/(Density*val_dist_ij);
+  //const su2double xiBulk = 1.0*total_viscosity/(Density*val_dist_ij);
+  const su2double xiBulk = 0.0;
```
I'm guessing these were corrections to the bulk viscosity.  Is that correct? What's our current understanding here?

trevilo replied:
Yes, that is correct.  I do not have a good understanding.  We should leave out all the bulk viscosity foo now as it is still experimental.

clarkpede replied:
Okay.  I'm tempted to make it a cfg option, but I'm unsure of what other options we would include besides a simple "Stokes hypothesis" option.  It doesn't seem like many compressible DNS or compressible LES researchers use anything different.

clarkpede commented on SU2_CFD/src/solver_direct_mean.cpp:
```diff
@@ -2273,23 +2280,55 @@ void CEulerSolver::Set_MPI_Solution_Gradient(CGeometry *geometry, CConfig *confi
+	  }
+	}
+
+
```
Is this a hack, or is it an improvement? Should it stay in the code?

trevilo replied:
This is intended to be an improvement.  The old code transforms the velocity gradient as if it were a vector not a tensor, which is wrong.  The new version is intended to fix that.  But, it hasn't been tested much, so it could still be buggy.

clarkpede commented on SU2_CFD/src/solver_direct_mean.cpp:
```diff
@@ -19184,7 +19681,7 @@ void CNSSolver::UpdateAverage(const su2double weight,
     /*-- Give Pk a different averaging time than the rest of the terms ---*/
 
-    const su2double new_Pk = production + (current_production - production)*weight*0.25;
+    const su2double new_Pk = production + (current_production - production)*weight*0.5; // for Cave=2
```
One of us needs to make this a cfg option.

trevilo replied:
Agreed.

clarkpede commented on SU2_CFD/src/variable_direct_mean.cpp:
```diff
@@ -837,32 +844,39 @@ bool CNSVariable::SetPrimVar(su2double eddy_visc, su2double turb_ke, CFluidModel
     
     RightVol = false;
-    
+    return RightVol;
```
What's going on here? I don't understand these changes.

trevilo replied:
Um, this is just hackery.  The default behavior is that when a problem (i.e., a non-physical value) is encountered, it overwrites the solution at the problem point with the old solution.  To the extent that this works (i.e., allows you to keep iterating toward a steady solution), it is fine for trying to converge a steady RANS solution.  In a hybrid run, it doesn't make sense.  But, my main objective wasn't to fix the behavior, it was just to not overwrite the problem point so that I could get information about it later.  But, like I said, it's just hackery and should be taken out.

More generally, we should talk about how to handle non-physical values in hybrid runs.  My take is that the code should just stop, write a solution file so we can see what the state is, and dump some info to the screen about the failure.  One can imagine trying to recover I guess, but it is a challenging task and I don't think it is worth it right now.

clarkpede replied:
Can you make a commit with the behavior you would like?

clarkpede commented on SU2_CFD/src/solver_structure.cpp:
```diff
@@ -2988,7 +2988,44 @@ void CSolver::SetAverages(CGeometry* geometry, CSolver** solver,
+	const su2double C_T = 6.0;
+	const su2double kol_time     = C_T*sqrt(nu/tdr_lim);
+	timescale = std::max(typical_time, kol_time);
+
```
There's an easier way to do this with the accessors for the "Typical" timescale and the kolmogorov timescale in `CTurbKEVariable`.  If we use the (previously computed) data, we'll avoid this duplication.

trevilo replied:
Yep, sounds good.  I'm in favor of avoiding duplication.

clarkpede replied:
Fixed.

clarkpede commented:
One open question is how to "validate" these changes.  The commit used to run the axisymmetric bump case diverges when applied to the low-Mach channel.  So we can't use the low-Mach channel to compare the baseline to any future changes.  I'm not sure how computationally expensive axisymmetric bump was, but we also didn't get great validation results.  So if something changes, it will be hard to tell if we "broke" something.

My proposal is that we use the low-Mach channel to compare the current state of pecos-dev (after the pecos-hybrid-channel-runs merge) with the current state of the PR.  Ideally, the low-Mach channel should work for both versions, without a decrease in accuracy.  Any other ideas for validating the code changes?

trevilo commented:
> My proposal is that we use the low-Mach channel to compare the current state of pecos-dev (after the pecos-hybrid-channel-runs merge) with the current state of the PR. Ideally, the low-Mach channel should work for both versions, without a decrease in accuracy. 

Yes, I agree.  So, I must have introduced a bug.  I doubt it will be terribly hard to find.  Worst case we can do a git bisect, but it might be obvious on code review.  I suggest I take this as an action item while you push forward on the high M channel.  But, if you want to do it, then I certainly won't get in the way.

clarkpede commented:
> So, I must have introduced a bug.

I wouldn't necessarily call it a "bug." It may be that the best parameters for a transonic run fail for a low-mach run.  But either way, we should find out what's causing the difference.

trevilo commented:
> I wouldn't necessarily call it a "bug." It may be that the best parameters for a transonic run fail for a low-mach run.

Fair enough, but I'd be shocked if it isn't a bug.

trevilo commented:
> The commit used to run the axisymmetric bump case diverges when applied to the low-Mach channel.

@clarkpede, just to confirm, the case that diverges is the low Mach channel that I put together, correct?  No changes to the config file or grid or anything?  If so, I'll investigate on that case.  If not, will you provide me with your inputs.  Thanks.

trevilo commented:
> The commit used to run the axisymmetric bump case diverges when applied to the low-Mach channel. 

One more question, just to help me set up my testing: how quickly does it diverge for you (in terms of number of time steps)?

clarkpede reviewed (commented):
When I ran the low-Mach channel case using the axisymmetric bump, I just used the same cfg file you gave me.  I don't think there were any modifications. The case diverged around  3400 iterations, but it looks like there were problems long before then.  I haven't had time to dive in and examine the results yet.

clarkpede committed:
Remove hacks to upwind dissipation

Previously, the code was modified to try using different blending values
to give different levels of upwinding to different transport equations.
It did not solve the problems at the time, and is removed here in favor
of simplicity.

trevilo commented:
Ok, sounds good.  I have a bisect going.  It is taking awhile b/c the queue is slow right now, but there aren't that many commits so we should know in the next day or two.

clarkpede committed:
Remove (unused) bulk viscosity correction

clarkpede committed:
Add config as parameter to res. tensor calls

Some of the tests failed to compile because they were not updated to
match the new code, where config is a parameter for the
SetResolutionTensor call.

clarkpede commented on SU2_CFD/src/numerics_direct_mean_hybrid.cpp:
```diff
@@ -345,7 +354,7 @@ void CAvgGrad_Hybrid::AddTauSGETJacobian(su2double *val_Mean_PrimVar,
 	}
 	tauSGET_mom[i][j] += -xi*mu[i][k]*nvec[j]*nvec[k];
-	tauSGET_mom[i][j] -= -xi*TWO3*mu[j][k]*nvec[i]*nvec[k];
+	//tauSGET_mom[i][j] -= -xi*TWO3*mu[j][k]*nvec[i]*nvec[k];
```
I don't think this change is correct, if we're still using the deviatoric velocity gradients.  Based on my derivations, I would expect the correct line to be:
```
tauSGET_mom[i][j] += -xi*mu[i][k]*nvec[j]*nvec[k]/3.0;
```

trevilo replied:
Yes, I agree the Jacobian of tauSGET is incorrect.  My derivation used the full velocity gradient, not the deviatoric part, and subtracted out the trace of tauSGET.  That is where the commented out term came from.

In any case, I get two extra contributions: the line you had and

`tauSGET_mom[i][j] -= -xi*mu[k][i]*nvec[j]*nvec[k]/3.0;`

Both lines were added in 1ea2891, and I have a channel case running now to test.

clarkpede committed:
Add a config option to control positivity of Pv2

clarkpede committed:
Add an option for production relaxation factor

clarkpede committed:
Limit hybrid average updates to model-split cases

This commit adds two changes to increase robustness of the code.  First,
an assert is added to ensure that the hybrid mediator exists before it
is called (when the code is executed from a debug version).  Second, the
averaging of hybrid variables only occurs when a model-split
hybridization is being used.

clarkpede committed:
Fix misleading v2f comments

Values passed to initialize v2f variables are primitive variables, not
conservative (density-weighted) variables.

clarkpede committed:
Refactor turbulent timescale calc in averaging

By using the right accessors, the kolmogorov and "typical" turbulent
timescales can be retrieved with no duplicate calc. Similarly, the
magnitude of the freestream velocity is precomputed, and does not need
to be recalculated.

trevilo commented on SU2_CFD/src/solver_direct_mean.cpp:
```diff
@@ -5315,6 +5478,27 @@ void CEulerSolver::Upwind_Residual(CGeometry *geometry, CSolver **solver_contain
+	}
+
+      }
+
```
git bisect plus some additional testing indicates that this nonsense is creating problems for the channel test case.  The goal of these changes was to increase the upwinding in regions where the resolution adequacy parameter is not "small".  This was in response to spurious oscillations that were observed at the boundary layer edge in the axisymmetric bump case that would sometimes become non-physical (negative density, temperature, etc).  However, it was only marginally successful in that case.  Given that it is causing problems for the channel case, these changes should not be included in the PR.  I will prepare a commit removing them.  We can investigate the effect on the axisymmetric bump separately.

trevilo committed:
Remove hacks to increase upwinding as a fcn of rM

clarkpede committed:
Update test values for v2f

Changing the Rf modification to not be the default has changed the
residuals once again.  I checked the covergence history, skin friction
profile, and pressure coefficient profile for a low-Mach ZPG flat plate.
The results are nearly identical. These differences seem to be
transient.

clarkpede committed:
Add config options for high-AR blending

trevilo commented:
Just fyi, the low M channel is still diverging, just taking longer.  So... I'll keep pushing on that and let you know when it is fixed.

trevilo committed:
Initialize default_fluct_stress_AR_params to NULL to avoid seg fault

clarkpede committed:
Bug fix for default AR blending params

clarkpede committed:
Add high-AR blending parameters to cfg template

trevilo committed:
Bug fix in tauSGET Jacobian to account for fact that tauSGET computed using the deviatoric part of the fluctuating velocity gradient (not full vel grad).

trevilo committed:
Merge branch 'pecos-hybrid-axibump-runs' of https://github.com/pecos-hybrid/SU2 into pecos-hybrid-axibump-runs

trevilo commented:
Well... I thought that 1ea2891 had the low M channel running stable, but not quite.  It diverged (with negative density) after 19161 time steps.

Prior to the divergence, the statistics looked pretty good, but the mean was clearly better previously:

[uprof.pdf](https://github.com/pecos-hybrid/SU2/files/3679113/uprof.pdf)

The "40FT (new)" results are from 161 steps before the negative density.  No sign of problems in the mean (or the instantaneous field at first glance).

So, the hunt continues.

trevilo commented:
The channel branch (pecos-hybrid-channel-testing) at 1fe82a5eacedce7dd5c8c519354cdcfb2378e025 does not show the instability (ran over 25000 time steps).  No insight yet into what causes the difference.

clarkpede committed:
Move upwind/central blending calculation

Previously, the upwind/central blending calculation for convective
fluxes was performed *before* the strains and vortices were calculated.
This meant that the initial blending calculation used unitialized
values.  This commit moves the calculation after the strain/vorticity
calculation.

(cherry picked from commit 29be76a3a931fb4241777a4a889e6612e932cba1)

clarkpede committed:
Comunicate average variables to halo nodes

(cherry picked from commit 3da76d1407c77e4f87ec27a3f1963b33b53f8386)

clarkpede committed:
Update restart loading to allow loading production

(cherry picked from commit 779cd53f4b33819cf49cf0d824acbad4d95e3e44)

clarkpede committed:
Remove the second set of SetRoe_Dissipation

In order to conserve variables, the same numerics must be applied in the
same way on both sides of a face/edge.  This means the "Roe
dissipation," or the hybrid central/upwind blending constant, must used
in the same way on both sides of an edge.

As the solver loops through the grid, it would previously pull the "Roe
Dissipation," compute a new (face) value, and then overwrite the old
value with the new (face) value in both nodes.  Since each cell has
multiple neighboring nodes, this resulted in the "Roe Dissipation" in
each cell being changed multiple times.  With this change, there was no
guarantee that the Roe Dissipation values used on one side of a face
would be the same as the Roe Dissipation values used on the other side
of the face.

The end result was that the system did *not* conserve mass.  Turning off
hybrid upwind/central blending fixed the mass conservation problem.

By removing the overwriting step, the original cell-specific values are
stored in each node. The only downside is that face-specific values of
the Roe dissipation must be computed manually as a post-processing step.
The solver does not record the values used at each face.

(cherry picked from commit 8e6e40482543db58dc9ab77d19d42784577bccb0)

clarkpede committed:
Fix unitialized variable use on halo nodes.

(cherry picked from commit 28a8f372fa31ec25de6e55a13128c35f855f40c9)

clarkpede committed:
Small change to make code less error-prone

In my testing, this change in point numbering did not introduce any
changes.  But it does make sure that loading the local data and sorting
it across processors are done in the same way, making the code less
error-prone.

clarkpede committed:
Initialize all new variables

These variables were previously uninitialized on the halo (periodic)
nodes. Even though these variables aren't used during the solve, they are
passed around during the output process (for restarts and solution
visualization).  This could be addressed by:

1. Fixing output so that it doesn't use halo nodes. This isn't really
what we want to do.
2. Zero-initializing the variables.  This is a poor programming practice,
since it makes a true zero indistinguishable from a zero-initialized
variables when it is later used.  Putting dummy data in a variable to
initialize it only hides problems from a static analysis tool.
3. Properly initializing the variables.

Number 3 is the long term goal.  Number 2 is what is done in this
commit.

(cherry picked from commit 074441abf11501e15da75818c3dcabcb57aac311)

clarkpede committed:
Changes made to make restarts consistent

Previously, restarted RANS solutions differed from a full run.  Two
changes are introduced in this commit to make single-MPI-task RANS
solutions consistent for both restart and non-restart solutions.

(cherry picked from commit c1e6d97aec641c247e27e50252565c340d3faab5)

clarkpede committed:
Fix bug in MPI communication of average solution

(cherry picked from commit 88372a15e56c9dcdedaedd0562babce3f2a7a029)

clarkpede committed:
Initialize all arrays used for hybrid model terms

Previously, some arrays were left uninitialized, leading to segfaults.

(cherry picked from commit c9c3406cf730fbb4a90e67cb080e9cf1dceb19f1)

clarkpede committed:
Make hybrid forcing standard

This is two sets of changes:

1. Remove deprecated "forcing stress" terms
2. Consolidate "ForcingVector" and "Force" member variables.  There were
   two variables before, and their roles overlapped.  This probably led
   to some bugs, but I didn't confirm it.  It definitely wasn't correct.

(cherry picked from commit bba231715f6cc88842456153314c6d8d12805aec)

clarkpede committed:
Add restart loading for mean hybrid force

(cherry picked from commit f2540481a482dc5154538392fdae87ff185f7d40)

clarkpede committed:
Fix initialization of resolution adequacy for restarts

(cherry picked from commit 45037c7d05d10392f184ae97cde91afb8ede9643)

clarkpede committed:
Fix inconsistency in production.

After the averages are updated, the value of production may have been
changed.  Previously, this value was not ever pushed back to the
turbulence solution nodes (CTurbKEVariable).  This meant that the values
used for output were lagged by a timestep.

By adding an extra overwrite to the turbulence postprocessing, the new
production values are pushed to the turbulence nodes after averaging
updates.

(cherry picked from commit 0cad7f0f12190312f1348a8f8071f91b1d3088af)

clarkpede committed:
Update test values

In k-based turbulence models, eddy viscosity depends on laminar
viscosity, which depends on temperature, which depends on turbulent
kinetic energy.  By changing the order in which these are computed at
the end of each iteration, the transient turbulent variables change.

The change in test values was very small, and this is not believed to be
a regression.

clarkpede committed:
Update restart variables in unit tests

This is due to two changes in the code:

+ Production is now required for loading a hybrid from RANS
+ Mean force has been added to the average solution variables that must
  be stored and retrieved.

clarkpede committed:
Change how hybrid restarts are handled

Now, instead of trying to guess your intention, if you try to load the
resolved turbulent stress from a production-only restart, you'll just
get an error.

clarkpede commented:
Since this is a lot of commits, I went through the list of changes and identified the big changes to the actual model:

+ Keeping div(u) from being added to production in hybrid mode if `config->GetBoolDivU_inTKEProduction() = false`
+ Apply `alpha (2 - alpha)` to SGS heat flux
+ Add eddy-viscosity realizability limit as an option.
+ New options to define the high-AR blending from the cfg file.
+ New options for the "production relaxation" constant.

The rest of the changes are pure coding changes, such as initializing variables with 0.

clarkpede committed:
Merge branch 'pecos-dev' into pecos-hybrid-axibump-runs

trevilo committed:
Revert mod to average production.

In 4a5d36b I modified the way the average production was computed to
be closer to what I thought CDP was doing, but this was based on a
misreading of CDP.  The original calculation was correct and leads to
much better results on the low Mach channel test case.

clarkpede committed:
Bugfix for variable density (e.g. compressible)

Since the lenghtscale calculation uses the dynamic viscosity and the
density, the lengthscale matrix must also be divided by density.

clarkpede committed:
Small bug fixes.
